`x:Bind` is resource dictionary is not supported

In WinUI and UWP, you can create a resource dictionary with a code-behind class, which then allows you to use x:Bind inside that resource dictionary. This works fine for WinUI when using Uno, but when I try this with a WASM head I get the following error:

`1>CSC : error UXAML0001: Unable to find x:Class on the top level element`

Does anyone know how to resolve this error? Or is it just impossible to use x:Bind from a Resource Dictionary for a WASM head?

As an example, I created a small project to demonstrate this.

App.xaml:

```
<Application
    x:Class="UnoResourceCodeBehind.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:UnoResourceCodeBehind">
	
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
                <!-- Add resource dictionaries here -->
                <local:ResourceDictionary1 xmlns="using:UnoResourceCodeBehind"/>
            </ResourceDictionary.MergedDictionaries>
            <!-- Add resources here -->
        </ResourceDictionary>
    </Application.Resources>

</Application>
```

MainPage.xaml:
```
<Page
    x:Class="UnoResourceCodeBehind.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:UnoResourceCodeBehind"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Grid>
        <local:MyControl Style="{StaticResource MyStyle}"/>
    </Grid>
</Page>
```

MyControl.cs:
```
using Microsoft.UI.Xaml.Controls;
using System;
using System.Collections.Generic;
using System.Text;

namespace UnoResourceCodeBehind
{
    internal class MyControl : Control
    {
        public string GetText() => "text from control";
    }
}
```

ResourceDictionary1.xaml:
```
<ResourceDictionary x:Class="UnoResourceCodeBehind.ResourceDictionary1"
                    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
                    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
                    xmlns:local="using:UnoResourceCodeBehind">

    <Style x:Name="MyStyle" TargetType="local:MyControl">
        <Setter Property="Template">
            <Setter.Value>
                <ControlTemplate TargetType="local:MyControl">
                    <Grid>
                        <TextBlock Text="{x:Bind GetText()}"/>
                    </Grid>
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>
    
</ResourceDictionary>
```

ResourceDictionary1.xaml.cs:
```
using Microsoft.UI.Xaml;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace UnoResourceCodeBehind
{
    partial class ResourceDictionary1 : ResourceDictionary
    {
        public ResourceDictionary1()
        {
            this.InitializeComponent();
        }
    }
}
```

This is the build error I get when building for WASM:

```
1>CSC : error UXAML0001: Unable to find x:Class on the top level element
1>C:\Users\user1\source\repos\UnoResourceCodeBehind\UnoResourceCodeBehind\UnoResourceCodeBehind.Shared\ResourceDictionary1.xaml(10,22,10,22): error UXAML0001: An error was found in Grid
1>CSC : error UXAML0001: Processing failed for file C:\Users\user1\source\repos\UnoResourceCodeBehind\UnoResourceCodeBehind\UnoResourceCodeBehind.Shared\ResourceDictionary1.xaml (Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: An error was found in Grid ---> System.Exception: Unable to find x:Class on the top level element
1>C:\Users\user1\source\repos\UnoResourceCodeBehind\UnoResourceCodeBehind\UnoResourceCodeBehind.Shared\MainPage.xaml.cs(27,18,27,37): error CS1061: 'MainPage' does not contain a definition for 'InitializeComponent' and no accessible extension method 'InitializeComponent' accepting a first argument of type 'MainPage' could be found (are you missing a using directive or an assembly reference?)
1>C:\Users\user1\source\repos\UnoResourceCodeBehind\UnoResourceCodeBehind\UnoResourceCodeBehind.Shared\ResourceDictionary1.xaml.cs(14,18,14,37): error CS1061: 'ResourceDictionary1' does not contain a definition for 'InitializeComponent' and no accessible extension method 'InitializeComponent' accepting a first argument of type 'ResourceDictionary1' could be found (are you missing a using directive or an assembly reference?)
1>C:\Users\user1\source\repos\UnoResourceCodeBehind\UnoResourceCodeBehind\UnoResourceCodeBehind.Shared\App.xaml.cs(26,18,26,37): error CS1061: 'App' does not contain a definition for 'InitializeComponent' and no accessible extension method 'InitializeComponent' accepting a first argument of type 'App' could be found (are you missing a using directive or an assembly reference?)
1>Done building project "UnoResourceCodeBehind.Wasm.csproj" -- FAILED.
```

My project is targeting .NET 7, btw.

_Originally posted by @andrewkdci in https://github.com/unoplatform/uno/discussions/13545_

--- a/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
+++ b/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
@@ -353,6 +353,7 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 			if (topLevelControl.Type.Name == "ResourceDictionary")
 			{
 				_isTopLevelDictionary = true;
+				_xClassName = FindClassName(topLevelControl);
 
 				using (TrySetDefaultBindMode(topLevelControl))
 				{

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ContentControl.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ContentControl.cs
@@ -187,6 +187,42 @@ namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Controls
 			Assert.IsTrue(comboBox.SelectedIndex == -1);
 		}
 
+
+		[TestMethod]
+		[RunsOnUIThread]
+		public async Task When_FindName_ContentControl_Without_ContentTemplate()
+		{
+			var sut = new ContentControl
+			{
+				Width = 100,
+				Height = 100,
+				Content = new TextBox()
+			};
+
+			WindowHelper.WindowContent = sut;
+			await WindowHelper.WaitForLoaded(sut);
+
+			Assert.IsNotNull(sut.FindName("ContentElement"));
+		}
+
+		[TestMethod]
+		[RunsOnUIThread]
+		public async Task When_FindName_ContentControl_With_ContentTemplate()
+		{
+			var sut = new ContentControl
+			{
+				Width = 100,
+				Height = 100,
+				Content = new TextBox(),
+				ContentTemplate = new DataTemplate(() => new TextBlock())
+			};
+
+			WindowHelper.WindowContent = sut;
+			await WindowHelper.WaitForLoaded(sut);
+
+			Assert.IsNull(sut.FindName("ContentElement"));
+		}
+
 		private class SignInViewModel
 		{
 			public string UserName { get; set; } = "Steve";

--- a/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ContentControl.cs
+++ b/src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ContentControl.cs
@@ -187,7 +187,7 @@ namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Controls
 			Assert.IsTrue(comboBox.SelectedIndex == -1);
 		}
 
-
+#if HAS_UNO
 		[TestMethod]
 		[RunsOnUIThread]
 		public async Task When_FindName_ContentControl_Without_ContentTemplate()
@@ -222,6 +222,7 @@ namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Controls
 
 			Assert.IsNull(sut.FindName("ContentElement"));
 		}
+#endif
 
 		private class SignInViewModel
 		{

--- a/src/Uno.UI.Tests/App/App.xaml
+++ b/src/Uno.UI.Tests/App/App.xaml
@@ -6,6 +6,7 @@
 			 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
 			 xmlns:views="using:Uno.UI.Tests.App.Views"
 			 xmlns:local="using:Uno.UI.Tests.App.Xaml"
+             xmlns:xbind="using:Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests.Controls"
 			 mc:Ignorable="not_win">
 	<Application.Resources>
 		<ResourceDictionary>
@@ -31,6 +32,7 @@
 				</ResourceDictionary>
 				<local:Subclassed_Dictionary />
 				<local:Subclassed_Dictionary_With_Property Test="Test123"/>
+                <xbind:XBind_ResourceDictionary />
 			</ResourceDictionary.MergedDictionaries>
 			<ResourceDictionary.ThemeDictionaries>
 				<ResourceDictionary x:Key="Light">

--- a/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary.xaml
+++ b/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary.xaml
@@ -1,0 +1,30 @@
+﻿<ResourceDictionary
+    x:Class="Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests.Controls.XBind_ResourceDictionary"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:controls="using:Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests.Controls">
+
+    <!--  Implicitly applied default style  -->
+    <Style BasedOn="{StaticResource DefaultxBindProjectTemplateStyle}" TargetType="controls:XBind_ResourceDictionary_Control" />
+
+    <Style x:Key="DefaultxBindProjectTemplateStyle" TargetType="controls:XBind_ResourceDictionary_Control">
+        <Style.Setters>
+            <Setter Property="Template">
+                <Setter.Value>
+                    <ControlTemplate TargetType="controls:XBind_ResourceDictionary_Control">
+                        <ContentControl
+                            x:Name="ContentContainer"
+                            HorizontalContentAlignment="Stretch"
+                            VerticalContentAlignment="Stretch">
+                            <ContentControl.ContentTemplate>
+                                <DataTemplate x:DataType="controls:XBind_ResourceDictionary_Control">
+                                    <Button Content="Hello" Loaded="{x:Bind Element_Loaded}" />
+                                </DataTemplate>
+                            </ContentControl.ContentTemplate>
+                        </ContentControl>
+                    </ControlTemplate>
+                </Setter.Value>
+            </Setter>
+        </Style.Setters>
+    </Style>
+</ResourceDictionary>

--- a/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary.xaml.cs
+++ b/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary.xaml.cs
@@ -1,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices.WindowsRuntime;
+using Windows.Foundation;
+using Windows.Foundation.Collections;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Data;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Navigation;
+
+namespace Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests.Controls;
+public sealed partial class XBind_ResourceDictionary : ResourceDictionary
+{
+	public XBind_ResourceDictionary()
+	{
+		this.InitializeComponent();
+	}
+}

--- a/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary_Control.cs
+++ b/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/XBind_ResourceDictionary_Control.cs
@@ -1,0 +1,30 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+
+namespace Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests.Controls;
+
+internal class XBind_ResourceDictionary_Control : Control
+{
+	/// <summary>
+	/// Creates a new instance of the <see cref="ProjectTemplate_xBind"/> class.
+	/// </summary>
+	public XBind_ResourceDictionary_Control()
+	{
+		this.DefaultStyleKey = typeof(XBind_ResourceDictionary_Control);
+
+		// Allows directly using this control as the x:DataType in the template.
+		this.DataContext = this;
+	}
+
+	public bool ElementLoadedInvoked { get; private set; }
+
+	public void Element_Loaded(object sender, RoutedEventArgs args)
+	{
+		ElementLoadedInvoked = true;
+	}
+}

--- a/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs
+++ b/src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs
@@ -1461,6 +1461,15 @@ namespace Uno.UI.Tests.Windows_UI_Xaml_Data.xBindTests
 			Assert.AreEqual("Updated2", SUT.tbDict2.Text);
 		}
 
+		[TestMethod]
+		public void When_XBind_In_ResourceDictionary()
+		{
+			var SUT = new XBind_ResourceDictionary_Control();
+			SUT.ForceLoaded();
+
+			Assert.IsTrue(SUT.ElementLoadedInvoked);
+		}
+
 		private async Task AssertIsNullAsync<T>(Func<T> getter, TimeSpan? timeout = null) where T : class
 		{
 			timeout ??= TimeSpan.FromSeconds(1);

--- a/src/Uno.UI/UI/Xaml/IFrameworkElement.cs
+++ b/src/Uno.UI/UI/Xaml/IFrameworkElement.cs
@@ -255,9 +255,17 @@ namespace Windows.UI.Xaml
 
 			// If element is a ContentControl with a view as Content, include the view and its children in the search,
 			// to better match Windows behaviour
-			var content =
-				(e as ContentControl)?.Content as IFrameworkElement ??
-				(e as Controls.Primitives.Popup)?.Child as IFrameworkElement;
+			IFrameworkElement content = null;
+			if (e is ContentControl contentControl &&
+				contentControl.Content is IFrameworkElement innerContent &&
+				contentControl.ContentTemplate is null) // Only include the Content view if there is no ContentTemplate.
+			{
+				content = innerContent;
+			}
+			else if (e is Controls.Primitives.Popup popup)
+			{
+				content = popup.Child as IFrameworkElement;
+			}
 
 			if (content != null)
 			{

